packages/teleterm/src/ui/DocumentTerminal/useDocumentTerminal.ts

```diff
@@ -48,10 +56,28 @@ async function initState(
   docsService: DocumentsService,
   doc: Doc
 ) {
+  const getClusterActualName = () => {
+    const cluster = ctx.clustersService.findCluster(clusterUri);
+    if (cluster) {
+      return cluster.actualName;
+    }
+
+    /*
+     When restoring the documents, we do not always have the leaf clusters already fetched.
+     In that case we can fall back to `clusterId` from a leaf cluster URI
+     (for a leaf cluster `clusterId` === `actualName`)
+    */
+    const parsed = routing.parseClusterUri(clusterUri);
+
+    if (!parsed?.params?.leafClusterId) {
+      throw new Error('Incorrect cluster URI');
```

Comment:
This error message could provide more information though. :D Maybe something that at this point a leaf cluster URI was expected but the URI doesn't seem to include a leaf cluster ID?